src/ui/public/visualize/visualize_legend.js

```diff
@@ -98,14 +104,28 @@ uiModules.get('kibana')
       ];
 
       function refresh() {
+        if (!$scope.renderbot) return;
         const vislibVis = $scope.renderbot.vislibVis;
 
         if ($scope.uiState.get('vis.legendOpen') == null && $scope.vis.params.addLegend != null) {
           $scope.open = $scope.vis.params.addLegend;
         }
 
-        $scope.labels = getLabels($scope.data, vislibVis.visConfigArgs.type);
-        $scope.getColor = colorPalette(_.pluck($scope.labels, 'label'), $scope.uiState.get('vis.colors'));
+        if (vislibVis.visConfigArgs.type === 'heatmap') {
+          const labels = vislibVis.getLegendLabels();
+          if (labels) {
+            $scope.labels = _.map(labels, label => { return { label: label }; });
+          } else {
+            $scope.labels = [{ label: 'loading ...' }];
+            $timeout(refresh, 100);
```

Comment:
the problem currently is that legend asks the chart to get the labels and colors, which only renders after the legend, thats why the legend is waiting until chart is ready .... an event or a promise would probably be better. what do you think ?